### PR TITLE
CORE-1282: support OAuth authorization callbacks in Sonora

### PR DESCRIPTION
**Pull request by slr71**
image
image

### CODE REVIEW

File: public/static/locales/en/oauth.json
Comment (slr71) [inline public/static/locales/en/oauth.json:8]: These error codes come directly from the OAuth specification. Having a unique message for each code will help us to troubleshoot these errors if they occur.

File: public/static/locales/en/oauth.json
Comment (slr71) [inline public/static/locales/en/oauth.json:11]: These error codes are specific to the DE:

- `no_auth_code_provided` means that the `code` query parameter was missing from the callback URL.
- `no_state_id_provided` means that the `state` query parameter was missing from the callback URL.
- `general_service_error` means that the API call to finish the authorization process failed. In this case, we can look at the `apps` and `terrain` logs for more information.

File: src/components/oauth/OAuthCodeHandler.js
Comment (slr71) [inline src/components/oauth/OAuthCodeHandler.js:38]: I used a general app listing URL as the default. That seemed to be the most reasonable thing to do if we can't determine the URL for HPC app listings.

File: src/components/oauth/OAuthCodeHandler.js
Comment (slr71) [inline src/components/oauth/OAuthCodeHandler.js:55]: Redirecting users here means that this component is specific to Tapis If we ever add support for other third-party APIs, we'll have to revisit this.

File: src/components/utils/error/ClientInfo.js
Comment (slr71) [inline src/components/utils/error/ClientInfo.js:4]: I extracted this from `ErrorHandler.js`. That seemed like the most reasonable thing to do to me. I'm curious to know what others think, though.

File: src/components/oauth/OAuthCodeHandler.js
Comment (slr71) [inline src/components/oauth/OAuthCodeHandler.js:22]: Importing `getListingPath` directly from `components/apps/utils.js` seemed reasonable to me, but I'm curious about what others think.

File: src/components/utils/error/ContactSupport.js
Comment (slr71) [inline src/components/utils/error/ContactSupport.js:3]: I also extracted this from `ErrorHandler.js`. Once again, this seemed to me to be the most reasonable thing to do, but I'm curious about what others think.

**Comment by slr71** [comment]
I'm not sure why the lint check is failing. I haven't been able to get the check to fail on my machine.

**Comment by slr71** [comment]
Still no clue why the lint check is failing. 🤷‍♀️

**Comment by slr71** [comment]
We figured out why the lint check was failing. There was a change that had been merged into master that wasn't in my initial PR to add `useEffect` on line 61 of `ErrorHandler.js`. I didn't see the changes locally because I initially didn't know that another PR had been merged and I wasn't merging in the changes from master. After I did the rebase, I probably could have found that `useEffect` call if I'd done the lint check locally after doing the rebase. I want to believe that I did, but I probably didn't. My only excuse would be that it was the end of the day and I was tired. 🤦‍♀️  At any rate, mystery solved. 😂

**Comment by sri2k1us** [comment]
LGTM. I recommend getting the messages reviewed by the 411 team to make sure the error messages are clear and easy to understand for non-tech users.

**Review by sri2k1us — APPROVED** [review]
:shipit:

**Comment by slr71** [comment]
Thanks for the review! I'll ping the 411 team to ask them to review the messages.

File: public/static/locales/en/oauth.json
Comment (ash3rz) [inline public/static/locales/en/oauth.json:3]: Just picking a random spot - I agree we should have 411 review the text.  I wouldn't be surprised if a user thought "remote system" meant their system and they started messing with permission settings on their computer or something like that 😄

**Review by ash3rz — APPROVED** [review]
LGTM and I like the refactors 👍 🎉 
As Sriram already said, having 411 review the text would be good 👍

File: src/components/oauth/OAuthCodeHandler.js
Comment (psarando) [inline src/components/oauth/OAuthCodeHandler.js:73]: As an alternative to using state to manage the app category query key, the query's `config` can use the `enabled` flag based on the presence of the user's profile ID, and this query key array can be used directly in the `queryKey`.

Here's an example where the `enabled` flag is based on having a user ID (though it's not used in the query key here):
https://github.com/cyverse-de/sonora/blob/3d8cb7640ef34c6192d8559f8d33f7fbbc48931f/src/pages/apps/%5BsystemId%5D/%5BappId%5D/edit.js#L34-L44

File: src/components/oauth/OAuthCodeHandler.js
Comment (psarando) [inline src/components/oauth/OAuthCodeHandler.js:96]: If we use `router.replace` here, then the browser's back button won't take the user back to the `oauth/callback` page from that `listingUrl`.

File: src/components/oauth/OAuthErrorHandler.js
Comment (psarando) [inline src/components/oauth/OAuthErrorHandler.js:56]: FYI, the `CardHeader` component accepts props `titleTypographyProps` and `subheaderTypographyProps`, since it's already wrapping the `title` and `subheader` in its own `Typography` components.

That means this text will be double-wrapped in `Typography` components (which probably isn't a big deal).

This code looks pretty tidy, though, so I'll leave it to personal preference, but here's an example where I'm using `titleTypographyProps`:
https://github.com/cyverse-de/sonora/blob/3d8cb7640ef34c6192d8559f8d33f7fbbc48931f/src/components/apps/AppStepDisplay.js#L58-L65

**Review by psarando — APPROVED** [review]
LGTM 👍 ✨ 

I just had a couple of minor alternate-style suggestions.

File: stories/OAuth.stories.js
Comment (slr71) [reply stories/OAuth.stories.js:12]: I really like this feature. It was really convenient for this case.

File: public/static/locales/en/oauth.json
Comment (slr71) [reply public/static/locales/en/oauth.json]: Thanks for catching that.

**Comment by slr71** [comment]
Thanks for the reviews! I'll wait for feedback from 411 before merging this PR.

**Comment by slr71** [comment]
I'm going to merge this now.

**Pull request merged**
